### PR TITLE
Ruby: add regex injection query

### PR DESCRIPTION
**Pull request by nickrolfe**

### CODE REVIEW

**Comment by nickrolfe** [comment]
I still need to add a missing QLDoc comment.

More worryingly, the DCA runs show that this query causes significant slowdowns. I'll have to investigate when I have time.

**Comment by nickrolfe** [comment]
As far as I can tell from DCA, the new query takes about the same amount of time as any other dataflow query.

File: ruby/ql/lib/codeql/ruby/frameworks/StandardLibrary.qll [hunk 1]
```diff
@@ -335,3 +336,18 @@ class ModuleEvalCallCodeExecution extends CodeExecution::Range, DataFlow::CallNo
 
   override DataFlow::Node getCode() { result = this.getArgument(0) }
 }
+
+/** Flow summary for `Regexp.escape`. */
```
Comment (hmac) [inline ruby/ql/lib/codeql/ruby/frameworks/StandardLibrary.qll:340]: Not one for this PR, but I wonder if we ought to split out `StandardLibrary.qll` soon? Maybe something like:

```
ql/lib/codeql/ruby/frameworks
├── core
│   ├── Kernel.qll
│   └── Object.qll
│   └── Regexp.qll
└── std-lib
    ├── Date.qll
    └── Open3.qll
```

File: ruby/ql/lib/codeql/ruby/frameworks/StandardLibrary.qll [hunk 2]
```diff
@@ -335,3 +336,18 @@ class ModuleEvalCallCodeExecution extends CodeExecution::Range, DataFlow::CallNo
...
+class RegexpEscapeSummary extends SummarizedCallable {
+  RegexpEscapeSummary() { this = "Regexp.escape" }
+
+  override MethodCall getACall() {
+    result = API::getTopLevelMember("Regexp").getAMethodCall("escape").asExpr().getExpr()
```
Comment (hmac) [inline ruby/ql/lib/codeql/ruby/frameworks/StandardLibrary.qll:345]: [`Regexp.quote`](https://ruby-doc.org/core-3.0.2/Regexp.html#method-c-quote) is an alias for `Regexp.escape`, so we could add that here too.

File: ruby/ql/lib/codeql/ruby/regexp/RegExpInjectionCustomizations.qll [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    RegExpLiteralAsSink() { this.asExpr().getExpr() instanceof RegExpLiteral }
+  }
+
+  /**
+   * The first argument of a call to `Regexp.new`, considered as a flow sink.
+   */
+  class ConstructedRegExpAsSink extends Sink {
+    ConstructedRegExpAsSink() {
+      this =
+        API::getTopLevelMember("Regexp").getAnInstantiation().(DataFlow::CallNode).getArgument(0)
```
Comment (hmac) [inline ruby/ql/lib/codeql/ruby/regexp/RegExpInjectionCustomizations.qll:56]: There are a few other ways I've found for creating regexes:
- [`Regexp.compile`](https://ruby-doc.org/core-3.0.2/Regexp.html#method-c-compile), which is an alias for `Regexp.new`
- [`Regexp.union`](https://ruby-doc.org/core-3.0.2/Regexp.html#method-c-union), which will create a regex matching a union of its arguments

**Review by hmac — DISMISSED** [review]
This looks great to me! I only have a couple of small suggestions for additional sinks/sanitizers.

**Comment by hmac** [comment]
@nickrolfe it looks like some consistency tests are failing now, which I assume is unrelated to this change. Not sure what we can do to fix them, but ping me if you get this green and I'll 👍  it.

**Comment by nickrolfe** [comment]
@hmac it's all green now!

**Review by hmac — APPROVED** [review]

**Pull request merged**